Add missing opcache return types for functions in spl

TysonAndre opened the pull request:
(excluding spl_autoload)

spl_object_id() is of the most interest to me,
since I frequently call it in an application.

This includes false/null types caused by wrong argument types and wrong argument
counts.

I can't rule out iterator_to_array returning null in spl_iterator_apply,
so leave MAY_BE_NULL in.

This targets php 7.4. I'm not sure if there are guidelines for what signatures get added to zend_func_info.c.

nikic commented:
I've cleaned up the code a bit in https://github.com/php/php-src/commit/5acedabfc04ddbb53d68a52a4ebcccca3090cbdc. That means iterator_count() and iterator_to_array() can't return null.

nikic commented on ext/opcache/Optimizer/zend_func_info.c:
```diff
@@ -1601,6 +1601,17 @@ static const func_info_t func_infos[] = {
+	F1("class_uses",							MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_STRING | MAY_BE_ARRAY_OF_STRING),
+	F0("iterator_apply",						MAY_BE_NULL | MAY_BE_FALSE | MAY_BE_LONG),
+	F0("iterator_count",						MAY_BE_NULL | MAY_BE_FALSE | MAY_BE_LONG),
+	F1("iterator_to_array",						MAY_BE_NULL | MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_ANY | MAY_BE_ARRAY_OF_ANY),
```
I'm wondering if the iterator_to_array result can contain references, e.g. when applied to generators? (Both in what is actually happening, and in what should be happening).

TysonAndre replied:
It looks like I can with https://www.php.net/manual/en/language.generators.syntax.php#control-structures.yield.references

```
php > $f = function &() use (&$x) { $x = -1; yield $x; };
php > foreach ($f() as &$y) { $y = 41; }
php > var_export($x);
41
php > $arr = iterator_to_array($f());
php > $arr[0] = 3;
php > var_export($x);
3
```

TysonAndre replied:
I don't see any examples of FN, FR, and FX being used, so I don't know which applies here

TysonAndre commented:
> I've cleaned up the code a bit in 5acedab. That means iterator_count() and iterator_to_array() can't return null.

This PR is targeting php 7.4, and that commit targets master?

nikic commented:
> This PR is targeting php 7.4, and that commit targets master?

The commit is in 7.4, GitHub just doesn't understand branches ;)

TysonAndre commented:
> I've cleaned up the code a bit in 5acedab. That means iterator_count() and iterator_to_array() can't return null.

Okay, I changed the signatures for those 2 functions

nikic commented on ext/opcache/Optimizer/zend_func_info.c:
```diff
@@ -1607,6 +1607,17 @@ static const func_info_t func_infos[] = {
+	F1("class_implements",						MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_STRING | MAY_BE_ARRAY_OF_STRING),
+	F1("class_parents",							MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_STRING | MAY_BE_ARRAY_OF_STRING),
+	F1("class_uses",							MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_STRING | MAY_BE_ARRAY_OF_STRING),
+	F0("iterator_apply",						MAY_BE_FALSE | MAY_BE_LONG),
```
```suggestion
	F0("iterator_apply",						MAY_BE_NULL | MAY_BE_LONG),
```

TysonAndre replied:
No, the signature I wrote should be correct, according to your earlier comments

> I've cleaned up the code a bit in 5acedab. That means iterator_count() and iterator_to_array() **can't return null.**

```php
php > var_export(iterator_count());
Warning: iterator_count() expects exactly 1 parameter, 0 given in php shell code on line 1
false
```

TysonAndre replied:
Oh, I changed the wrong function

nikic commented on ext/opcache/Optimizer/zend_func_info.c:
```diff
@@ -1607,6 +1607,17 @@ static const func_info_t func_infos[] = {
+	F1("class_uses",							MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_STRING | MAY_BE_ARRAY_OF_STRING),
+	F0("iterator_apply",						MAY_BE_FALSE | MAY_BE_LONG),
+	F0("iterator_count",						MAY_BE_FALSE | MAY_BE_LONG),
+	F1("iterator_to_array",						MAY_BE_NULL | MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_ANY | MAY_BE_ARRAY_OF_REF | MAY_BE_ARRAY_OF_ANY),
```
```suggestion
	F1("iterator_to_array",						MAY_BE_FALSE | MAY_BE_ARRAY | MAY_BE_ARRAY_KEY_ANY | MAY_BE_ARRAY_OF_REF | MAY_BE_ARRAY_OF_ANY),
```

TysonAndre commented:
Review comments have been addressed and the suggested changes were merged.

nikic commented:
Merged as https://github.com/php/php-src/commit/94e2f25f076734a38b1c046e06459d404a7650cf, thanks!
